waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> makeError nextToken (TokenSet.fromKind kind)
+
+-- | Parse "EOF" rule. If there are no non-trivia tokens in the remaining input,
+-- this parser succeeds without making any nodes. Otherwise, it consumes all
+-- remaining input and reports an error at the first non-trivia token.
+parseEOF :: ParserM ()
+parseEOF =
+  (consumeWhitespace >> peek) >>= \case
+    Just nextToken -> do
+      startOffset <- gets pstateNextOffset
+      advance
+      let firstTokenKind = tokenKind nextToken
+      -- TODO: consider reporting all kinds here? not sure
+      _tokenKinds <- collectUntilEOF
+      throwError $
+        ParseError $
+          Unexpected
+            (Span startOffset (startOffset + tokenWidth nextToken))
+            firstTokenKind
+            TokenSet.fromEOF
+      where
+        collectUntilEOF :: ParserM [TokenKind]
+        collectUntilEOF =
+          peek >>= \case
+            Just next -> advance >> ((tokenKind next :) <$> collectUntilEOF)
+            Nothing -> pure []
+    Nothing -> do
+      return ()
+
+-- | Parse "As" rule. Identical to "Primitive", but uses the label as the kind
+-- for the syntax node when it succeeds.
+parseAs :: SyntaxKind -> TokenKind -> ParserM ()
+parseAs label kind =
+  (consumeWhitespace >> peek) >>= \case
+    Just nextToken | tokenKind nextToken == kind -> do
+      advance
+      let newNode =
+            SyntaxNode
+              { snodeKind = label,
+                snodeWidth = tokenWidth nextToken,
+                snodeChildren = []
+              }
+      pushNode newNode
+    nextToken -> makeError nextToken (TokenSet.fromKind kind)
+
+-- | Parse "Group" rule. Runs the inner parser. When it succeeds, it puts the
+-- outputted nodes from that parser under a parent node with the label. When it
+-- fails, it recovers from the error by catching the error or unwinding.
+parseGroup :: SyntaxKind -> GrammarRule -> ParserM ()
+parseGroup label inner = do
+  let pushGroupNode children =
+        let width = sum $ map snodeWidth children
+         in pushNode $
+              SyntaxNode
+                { snodeKind = label,
+                  snodeWidth = width,
+                  snodeChildren = children
+                }
+
+  consumeWhitespace
+  siblingNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = []})
+  result <- tryError $ doParse inner
+  childNodes <- gets pstateNodes
+  modify (\s -> s {pstateNodes = siblingNodes})
+  case result of
+    Nothing -> do
+      pushGroupNode childNodes
+    Just err -> do
+      case err of
+        Unwind kind -> do
+          -- When unwinding, check if this is the level where we have the
+          -- following token. If it is, then we can get rid of the unwind.
+          --
+          -- Otherwise, we re-throw the unwind
+
+          followingStack <- gets pstateFollowing
```

Comment:
I am trying to wrap my head around what "following" tokens represent.
Looking at code, there is only one place where we `pushFollowing`, which is in Chain -> before we try to parse the first rule in the chain, we push the first tokens of the second rule. Ok, so let's say we go down the first rule, and then there is an error.
If there is no Group, it will go all the way to the top and end parsing, since only Group can do recovery, right? Ok, but there will be Group very likely, so Group catches the parsing error, and if the token that it errored on (errorToken let's call it) is among the following tokens (but not the first one), it will throw Unwind, to push this to the Group above.
If it is however among the first set of expected following tokens, we do nothing, right? That way group just ends and we continue on the next member of the Chain?

I find it pretty hard to follow all this, I guess due to all the state and maybe some assumptions? Are we assuming Group is part of the Chain? What if it is not? Am I talking gibberish now :D?

I also wonder how do we know this is a good way to do error recovery? But I would probably be better able to evaluate that if I better understood what is happening.

I would suggest two things:
1. Refactoring this code as much as possible to make it very easy to understand. It is not so bad at the moment, has some nice names and comments, but we could probably push it further. For example, it would make sense to first handle ParseError case, not Unwind case, since there is not much sense in reader trying to understand Unwind handling until they first saw how it is created from ParseError. Then, we could probably use `followingTokenStack` instead of `followingStack` -> there are some other places where `Token` word could be added for clarity. Maybe some additional reorganization of logic.
2. Could you "paint a picture" of how this works? Maybe an specific example that  you would go through step by step and visualize what happens here? We can do this together for start, until we get to a point where you can explain to me, and then we can look into putting it down in the comments somehow.